content/blog/integrating-python-3-with-fips-enabled-openssl-3-1.md

```diff
@@ -0,0 +1,149 @@
+
+> `cd openssl-3.1.0`\
+> `Perl Configure VC-WIN64A --prefix=C:\SSLout\DLL\x64\Release --openssldir=c:\SSLout\SSL enable-fips`\
+> `nmake`\
+> `nmake install_sw`\
+> `nmake install_ssldirs`\
+> `nmake install_docs`\
+> `nmake install_fips`
+
+The OpenSSL 3.1 binaries are generated be in the `C:\SSLout\DLL\x64\Release\bin`.  
+
+This is shown as below
+
+![](/img/openssl_directory_structure.jpg)
+
+Step 7 Unzip Python in Windows to `C:\work\Python311` and go into `C:\work\Python-3.11.2\PCBuild`.  
+
+Run `get_externals.bat` file.  This will fetch the dependencies from internet into externals directory. Now create `openssl-bin-3.1.0` directory under `externals` directory as below.
+
+Step 8 Under `externals` directory, create `amd64` depending on the CPU architecture and 
+copy the as files to it as below.
+
+> copy `C:\SSLout\DLL\x64\Release\lib\libcrypto.lib` and `libssl.lib` to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\bin\*.*` to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\include` directory to `openssl-bin-3.1.0\amd64`\
+> copy `C:\SSLout\DLL\x64\Release\include\openssl\applink.c` to `openssl-bin-3.1.0\amd64\include`\
+> copy `C:\SSLout\DLL\x64\Release\lib\ossl-modules\*.*` to `openssl-bin-3.1.0\amd64`
+
+Step 9 Modify `PCbuild/openssl.props` as shown below:
+
+![](/img/openssl_settings.jpg)
+
+Step 10 Open `PCbuild/python.props` and change the entries as shown below:
+
+![](/img/openssl_settings2.jpg)
+
+Step 11 Open Python Solution under `PCbuild/pcbuild.sln` in VS 2015.
+Change the Link settings of `_hashlib` and `_ssl` projects under Python as shown below:
+
+![](/img/openssl_vs_settings.jpg)
+
+Step 12 Now, build `_hashlib.pyd` and `_ssl.pyd` in VS 2015.
+Step 13 Copy these built .pyd files from `\PCbuild\amd64\` to a Python binary installation directory `C:\python311\DLLs`.
+Step 14 Start Python and use these commands to check the OpenSSL 3.1 version.
+
+![](/img/openssl-after.jpg)
+
+Step 15 Now, to run Python with FIPS enabled OpenSSL, create `openssl.cnf` and `fipsmodule.cnf` file using the below content in `C:\Python311` directory. By using this config file, FIPS will be enabled by default.
+
+> #### openssl.cnf:
+>
+> config_diagnostics = 1\
+> openssl_conf = openssl_init\
+> .include .\fipsmodule.cnf\
+> \[openssl_init]
+> providers = provider_sect\
+> alg_section = algorithm_sect\
+> \[provider_sect]\
+> fips = fips_sect\
+> legacy = legacy_sect\
+> base = base_sect\
+> default = default_sect\
+> \[base_sect]\
+> activate = 1\
+> \[legacy_sect]\
+> activate = 1\
+> \[default_sect]\
+> activate = 1\
+> \[algorithm_sect]\
+> default_properties = fips=yes
+>
+> #### fipsmodule.cnf:
+>
+> \[fips_sect]\
+> activate = 1\
+> conditional-errors = 1\
+> security-checks = 1\
+> module-mac = D4:64:00:E3:CE:34:EE:CE:58:32:12:08:21:6D:64:FD:E3:A6:D4:F0:E6:38:3D:2C:0C:40:1B:50:C8:8F:39:A3
+
+Step 16 Now open command windows as Administrator and execute 
```

Comment:
Recommend rephrasing Step 16 to:

**Step 16:** Now, open the command window as an Administrator and execute the following command